service/src/main/kotlin/fi/espoo/evaka/messaging/MessageQueries.kt

```diff
@@ -1100,3 +1119,168 @@ fun Database.Read.getArchiveFolderId(accountId: MessageAccountId): MessageThread
+FROM message m
+WHERE sender_id = :accountId AND content_id = :contentId
+"""
+            )
+            .bind("accountId", accountId)
+            .bind("contentId", contentId)
+            .mapTo<MessageToUndo>()
+            .toList()
+
+    if (messagesToUndo.isEmpty()) {
+        throw BadRequest("No messages found with contentId $contentId")
+    }
+
+    if (
+        messagesToUndo.any { it.sentAt.plusSeconds(MESSAGE_UNDO_WINDOW_IN_SECONDS).isBefore(now) }
+    ) {
+        throw BadRequest(
+            "Messages older than $MESSAGE_UNDO_WINDOW_IN_SECONDS seconds cannot be undone"
+        )
+    }
+
+    if (messagesToUndo.any { !it.onlyMessageInThread }) {
+        throw BadRequest("Cannot undo message threads that have replies")
+    }
+
+    val draftContent =
+        this.createQuery(
+                """
+SELECT DISTINCT t.title, t.message_type AS type, t.urgent, c.content, '{}'::text[] AS recipient_ids, '{}'::text[] AS recipient_names
+FROM message_content c
+JOIN message m ON c.id = m.content_id
+JOIN message_thread t ON m.thread_id = t.id
+WHERE c.id = :contentId
+"""
+            )
+            .bind("contentId", contentId)
+            .mapTo<UpdatableDraftContent>()
+            .findOne()
+            .orElseThrow { error("Multiple draft contents found") }
+
+    this.deleteMessages(messagesToUndo)
+
+    val draftId = this.initDraft(accountId)
+    this.updateDraft(accountId, draftId, draftContent)
+    return draftId
+}
+
+fun Database.Transaction.deleteMessages(messages: List<MessageToUndo>) {
+    this.createUpdate(
+            """
+WITH deleted_message AS (
+    DELETE FROM message WHERE id = ANY(:messageIds)
+), deleted_recipients AS (
+    DELETE FROM message_recipients WHERE message_id = ANY(:messageIds)
+), deleted_thread AS (
+    DELETE FROM message_thread WHERE id = ANY(:threadIds)
+), deleted_thread_children AS (
+    DELETE FROM message_thread_children WHERE thread_id = ANY(:threadIds)
+)
+DELETE FROM message_content WHERE id = ANY(:contentIds)
+"""
+        )
+        .bind("messageIds", messages.map { it.messageId }.distinct())
+        .bind(
+            "threadIds",
+            messages.filter { it.onlyMessageInThread }.map { it.threadId }.distinct()
+        )
+        .bind("contentIds", messages.map { it.contentId }.distinct())
+        .execute()
+}
+
+fun Database.Transaction.updateThreadParticipants(
+    threadId: MessageThreadId,
+    senderId: MessageAccountId
+) {
+    this.createUpdate(
+            """
+UPDATE message_thread_participant p SET last_message_timestamp = any_message.sent_at, last_sent_timestamp = sent_message.sent_at
+FROM (
+    SELECT MAX(sent_at) AS sent_at
+    FROM message m
+    WHERE m.thread_id = :threadId AND m.sender_id = :senderId
+) sent_message
+JOIN (
+    SELECT MAX(sent_at) AS sent_at
+    FROM message m
+    WHERE m.thread_id = :threadId
+) any_message ON true
```

Comment:
The logic here doesn't work if there are messages in the thread that are not visible to `senderId`. You have to filter using the `message_recipients` table, too.

This made me think, would it be better to only update `last_sent_timestamp` and `last_message_timestamp` only for the sender when a message is sent, and then add an async job to update the `last_message_timestamp` for the recipients after the undo period has expired. Then the thread would be bumped to the top of inbox only after the recipient actually sees the message.

The async job could include the sent message id and only do its job if that message still exist, i.e. sending it hasn't been undone.

Reply:
Furthermore, I would personally find the query better in this form:

```
UPDATE ... 
SET
    x = (SELECT MAX(...) ...),
    y = (SELECT MAX(...) ...)
WHERE z
```

This would be easier for a human to parse (in my opinion) and as a bonus, it would be an error if the subqueries returned more than one row. This would minimize the risk that someone in the future changes the query so that it would affect the wrong set of rows.

If you decide to keep it such that it still has a `FROM` clause, I'd like to point out that `FROM a JOIN b ON true` is equivalent to `FROM a, b`.